File: app_programing/app_programing/samples/railbook/app/channels/message_channel.rb
class MessageChannel < ApplicationCable::Channel
  def subscribed
    # stream_from 'general'
    stream_from params[:topic]
  end

  def unsubscribed

  end

  def sendMessage(data)
    # ActionCable.server.broadcast 'general', { name: data['name'], body: data['body'] }
    # Message.create topic: 'general', name: data['name'], body: data['body']

    # トピックを分割する場合
    ActionCable.server.broadcast data['topic'],
      { name: data['name'], body: data['body'] }
    Message.create(topic: data['topic'], name: data['name'], body: data['body'])
  end
end

File: app_programing/app_programing/samples/railbook/app/javascript/channels/message_channel.js
import consumer from "channels/consumer"

const btn = document.querySelector('#btn');
const body = document.querySelector('#body');
const name = document.querySelector('#name');
const messages = document.getElementById('messages');
const paths = location.href.split('/');
const topic = paths[paths.length - 1];

// const app = consumer.subscriptions.create('MessageChannel', {
//   received(data) {
//     const p = document.createElement('p');
//     p.textContent = `${data.name}：${data.body}`;
//     messages.prepend(p);
//   },
//   sendMessage(name, msg) {
//     return this.perform('sendMessage', { name: name, body: msg });
//   }
// });


// トピックを分割する場合
const app = consumer.subscriptions.create(
  { channel: 'MessageChannel', topic }, {
  received(data) {
    const p = document.createElement('p');
    p.textContent = `${data.name}：${data.body}`;
    messages.prepend(p);
  },
  sendMessage(topic, name, body) {
    return this.perform('sendMessage', { topic, name, body });
  }
});

if (btn) {
  btn.addEventListener('click', function(e) {
    e.preventDefault();
    // app.sendMessage(name.value, body.value);
    app.sendMessage(topic, name.value, body.value);
    body.value = '';
  }, false);
}
